src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -0,0 +1,269 @@
+    PublicKey const& identity)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+    return ephemeral_.erase (identity);
+}
+
+bool
+ValidatorList::insertPermanentKey (
+    RippleAddress const& identity,
+    std::string const& comment)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    auto const publicKey = asPublicKey (identity);
+
+    if (ephemeral_.find (publicKey) != ephemeral_.end())
+    {
+        JLOG (j_.error) <<
+            toBase58 (TokenType::TOKEN_NODE_PUBLIC, publicKey) <<
+            ": permanent key exists in ephemeral table!";
+        return false;
+    }
+
+    return permanent_.emplace (publicKey, comment).second;
+}
+
+bool
+ValidatorList::removePermanentKey (
+    RippleAddress const& identity)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+    return permanent_.erase (asPublicKey (identity));
+}
+
+void
+ValidatorList::for_each (
+    std::function<void(PublicKey const&, std::string const&, bool)> func) const
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    for (auto const& v : permanent_)
+        func (v.first, v.second, false);
+    for (auto const& v : ephemeral_)
+        func (v.first, v.second, true);
+}
+
+//------------------------------------------------------------------------------
+static
+std::size_t
+loadValidators (
+    ValidatorList *vl,
+    std::vector<std::string> const& entries,
+    beast::Journal& j)
+{
+    static boost::regex const re (
+        "[[:space:]]*"            // skip leading whitespace
+        "([[:alnum:]]+)"          // node identity
+        "(?:"                     // begin optional comment block
+        "[[:space:]]+"            // (skip any amount of whitespace)
+        "(.*[^[:space:]]+)"       // the comment (at least non-whitespace)
+        "[[:space:]]*"            // (skip any amount of whitespace)
+        ")?"                      // end optional comment block
+    );
+
+    std::size_t count = 0;
+
+    for (auto const& n : entries)
+    {
+        boost::smatch match;
+
+        if (!boost::regex_match (n, match, re))
+        {
+            JLOG (j.error) <<
+                "Malformed entry: '" << n << "'";
+            continue;
+        }
+
+        auto const id = parseBase58<PublicKey>(
+            TokenType::TOKEN_NODE_PUBLIC, match[1]);
+
+        if (!id)
+        {
+            JLOG (j.error) <<
+                "Invalid node identity: " << match[1];
+            continue;
+        }
+
+        auto const ra = RippleAddress::createNodePublic (match[1]);
+
+        if (vl->member (ra))
+        {
+            JLOG (j.warning) <<
+                "Duplicate node identity: " << match[1];
+            continue;
+        }
+
+        if (vl->insertPermanentKey(ra, trim_whitespace (match[2])))
+            count++;
+    }
+
+    return count;
+}
+
+//------------------------------------------------------------------------------
+std::unique_ptr<ValidatorList>
+make_ValidatorList (Config const& config, beast::Journal j)
```

Comment:
This is bad form. We typically use a factory free function starting with `make_` when the implementation is hidden behind an abstract interface. And that free function never implements any business logic. This function however, implements business logic. It looks like it can't report a failure to the caller. Which makes sense, because the `ValidatorList` should always be constructible.

I suggest that `ValidatorList` offer a two-phase construction interface. First, call the constructor which does not do any loading. Then, the caller must call a member function like `loadFromFile`, and this will contain the logic currently in `make_ValidatorList`. The function `loadValidators` should obviously be a member function and not a file-scoped free function. It might be possible to reduce the public interface of `ValidatorList` after that (for example, by not offering a public method to insert new items).